src/libs/tools/midi.c

```diff
@@ -499,33 +557,57 @@ void midi_open_devices(dt_lib_module_t *self)
 
       if(error < 0)
       {
-        dt_print(DT_DEBUG_ALWAYS, "[midi_open_devices] ERROR opening midi device '%s' via '%s'\n",
+        dt_print(DT_DEBUG_ALWAYS, "[_midi_open_devices] ERROR opening midi device '%s' via '%s'\n",
                  info->name, info->interf);
         continue;
       }
       else
       {
-        dt_print(DT_DEBUG_INPUT, "[midi_open_devices] opened midi device '%s' via '%s' as midi%d\n",
+        dt_print(DT_DEBUG_INPUT, "[_midi_open_devices] opened midi device '%s' via '%s' as midi%d\n",
                  info->name, info->interf, dev);
       }
 
-      midi_device *midi = (midi_device *)g_malloc0(sizeof(midi_device));
+      dt_midi_device_t *midi = (dt_midi_device_t *)g_malloc0(sizeof(dt_midi_device_t));
 
       midi->id          = id + dev;
       midi->info        = info;
       midi->portmidi_in = stream_in;
-
-      midi->behringer = strstr(info->name, "X-TOUCH MINI") ? 'M'
-                      : strstr(info->name, "BCR2000"     ) ? 'R'
-                      : strstr(info->name, "BCF2000"     ) ? 'F'
-                      : 0 /* (X)-Touch (C)ompact/(E)tended/(O)ne */;
-
       midi->encoding    = encoding;
-      midi->num_knobs   = midi->behringer == 'M' ? 18 : midi->behringer ? 32 : 0;
-      midi->first_knob  = midi->behringer == 'M' ?  1 : 0;
-      midi->num_keys    = midi->behringer == 'M' ? 16 : num_keys;
-      midi->first_key   = midi->behringer == 'M' ?  8 : 0;
-      midi->first_light = 0;
+      midi->num_keys    = num_keys;
+
+      if(strstr(info->name, "X-TOUCH MINI"))
```

Comment:
Should we start thinking about adding an external JSON description for those definitions?

I see also in the code some check like `midi->behringer == 'M'`, so maybe we would want to have some boolean in the JSON to describe some specific supported feature?  To make the code more generic and so replace the check by:

```
   if(midi->has_such_feature)
   {
```

?

Reply:
> Should we start thinking about adding an external JSON description for those definitions

Of course I have considered this; I had many plans. :-)

This module basically provides support for "generic midi + several behringer extensions". 4 relatively affordable models are supported in a more or less hardwired way. Extrapolating from the way these machines have bolted additional features on top of the "standard" midi protocol would quickly lead to an overengineered solution that would be used by nobody (zero people) and receive even less testing and feedback. Even the newer behringer modules use different extensions. Generic support for any crazy midi machine would be ... crazy. It would have to be implemented via lua callbacks. Which would be feasible if we had enough data to define the required lua interface, which we have not.

(btw one device I would love to implement support for if somebody donated one is the Midi Fighter Twister (-: )

What at this point we _could_ put in a JSON file is the definitions of a lot of "standard" midi machines that use relative encoding (to say which encoding they use and how many keys and knobs they have and maybe if the keys have lights). https://darktable-org.github.io/dtdocs/en/special-topics/midi-device-support/ asks for feedback/input for such devices but the response has been limited to a handful.

But a bigger problem is that these devices are often extremely configurable, so the _default_ factory encoding, channel etc settings may have been changed by the users and presumably should not be overwritten. There _has_ been feedback about dt too aggressively wanting to engage with connected midi devices, so it seems people prefer to indicate themselves if they want to use one. Which they would do in "preference/misc/order or exclude midi devices". The standard settings to be used for some known devices are in the dtdocs page, but this could be turned into a reference list.